### PR TITLE
Settings editors  

### PR DESCRIPTION
**Pull request by 3nids**
Finally the cool stuff :)

* API
  * settings: Removal of QgsSettingsEntryByValue, only use by reference instead
  * old QgsSettingsTreeWidget has been renamed to QgsSettingsTreeWidgetOld (in app)
  * new classes:
    * QgsSettingsTreeWidget and QgsSettingsTreeModel: tree widget + model to show and edit the settings
    * QgsSettingsEditorWidgetWrapper (pure virtual): a base wrapper class to create and handle a setting editor widget
    * QgsSettingsEditorWidgetWrapperTemplate(pure virtual): the base class for wrapper of settings declared in c++ API (using typed methdos)
    * classes for settings editors: QgsSettingsStringEditorWidgetWrapper, …
    * QgsSettingsEditorWidgetRegistry: a registry of the setting editor wrappers

* UX (in the advanced options):
  * we have the choice of using the old or the new tree until the majority of settings are merged, choice is remembered
  * a "hidden" setting allows to bypass the warning before showing the settings tree
  * settings have dedicated editors (spin box, checkbox, colorbutton, …)
  * the modification of settings is now only achieved when apply is pressed
  * settings not being set yet are in italic, edited settings in red

To come in a next PR (not yet done)
  * [ ] respect constraints of the setting (min/max for int, transparency for colors, …)
  * [ ] tests (will come first)
  * [ ] inform when it's not possible to show and/or edit the setting value (lists and maps for now)
  * [x] ~~filtering~~
  * [ ] reset to defaultValue
  * [ ] delete named node item (e.g. a connection entry)
  * [ ] add missing editor: enums, 64bits integer, list, maps
  * [ ] context menu to delete, reset to original, copy/paste

### CODE REVIEW

**Comment by 3nids** [comment]
Screen Recording 2023-04-23 at 07 21 16

File: src/gui/settings/qgssettingstreemodel.h [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    //! Constructor
+    QgsSettingsTreeModel( QgsSettingsTreeNode *rootNode = nullptr, QObject *parent = nullptr );
+
+    ~QgsSettingsTreeModel();
+
+    //! Apply pending changes in the model to the corresponding settings
+    void applyChanges();
+
+    /**
+     * Returns settings tree node for given index. Returns root node for invalid index.
```
Comment (nirvn) [inline src/gui/settings/qgssettingstreemodel.h:186]:
```suggestion
     * Returns settings tree node for given \a index or the root node if the index is invalid.
```

File: src/app/options/qgsadvancedoptions.cpp [hunk 2]
```diff
@@ -14,26 +14,41 @@
...
+
   layout()->setContentsMargins( 0, 0, 0, 0 );
 
-  connect( mAdvancedSettingsEnableButton, &QPushButton::clicked, this, [ = ]
+  if ( !settingsShowWarning->value() )
   {
-    mAdvancedSettingsEditor->show();
     mAdvancedSettingsWarning->hide();
-  } );
+    mGroupBox->layout()->addWidget( createSettingsTreeWidget() );
```
[the remark this answers is not in the document]
Comment (3nids) [reply src/app/options/qgsadvancedoptions.cpp:41]: why you reset the setting value here, nothing changed.
`settingShowWarning` is a non GUI exposed setting, so setting it to false will show directly the type of tree you want.

**Comment by troopa81** [comment]
Nice work :+1: 

Just on UI side of things:
- Could we change the color of modified settings? I would say "red" means error for most of users
- I'm wondering if it would'nt be better/possible to display directly the checkbox, instead of true/false and then the checkbox when edit

**Comment by 3nids** [comment]
Thanks for the feedback @troopa81 

I've implemented the checkbox for the booleans and used the palette colors (see screenshots).
Also, filtering is now done.

image


image

**Comment by troopa81** [comment]
Thanks @3nids , it looks better :+1:

**Review by nirvn — APPROVED** [review]

**Pull request merged**